File: src/c/break_even_pid.c
#define __USE_MINGW_ANSI_STDIO 1
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct {
    long double kp;
    long double ki;
    long double kd;
    long double integral;
    long double prev_error;
} PIDController;

long double calculate_pid(PIDController *pid, long double setpoint, long double pv, long double dt) {
    long double error = setpoint - pv;
    pid->integral += error * dt;
    long double derivative = (error - pid->prev_error) / dt;
    long double output = pid->kp * error + pid->ki * pid->integral + pid->kd * derivative;
    pid->prev_error = error;
    return output;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <target_output>\n", argv[0]);
        return 1;
    }

    char *endptr;
    long double target = strtold(argv[1], &endptr);
    if (argv[1] == endptr) {
        fprintf(stderr, "Invalid input: %s\n", argv[1]);
        return 1;
    }

    // System: y = 10*u + 1. System gain is 10.
    // To stabilize, Kp should be small. 
    PIDController pid = {0.01L, 0.005L, 0.001L, 0.000L, 0.000L}; 
    long double current_u = 0.0L;
    long double dt = 0.1L;

    printf("Simulation for Target: %Lf (100m units)\n", target);
    printf("Step, Control_Input(u), System_Output(y), Error\n");

    for (int i = 0; i < 50; i++) {
        long double y = current_u * 10.0L + 1.0L;
        long double error = target - y;

        printf("%d, %15.6Lf, %15.6Lf, %15.6Lf\n", i, current_u, y, error);

        if (fabsl(error) < 1e-9L) break;

        long double delta_u = calculate_pid(&pid, target, y, dt);
        current_u += delta_u;
    }

    return 0;
}

File: src/go/break_even_pid.go
package main

import (
	"fmt"
	"math"
	"sync"
)

// PIDController represents a PID controller using float64 (high precision)
type PIDController struct {
	Kp, Ki, Kd float64
	Integral   float64
	PrevError  float64
}

func (pid *PIDController) Calculate(setpoint, pv, dt float64) float64 {
	error := setpoint - pv
	pid.Integral += error * dt
	derivative := (error - pid.PrevError) / dt
	output := pid.Kp*error + pid.Ki*pid.Integral + pid.Kd*derivative
	pid.PrevError = error
	return output
}

// System simulates the "Grand Strategy Map" logic: output = input * 10 + 1
func System(input float64) float64 {
	return input*10.0 + 1.0
}

// SimulationTask represents a single competitive bidding simulation
type SimulationTask struct {
	BidID  int
	Target float64 // Break-even point in 100m units
}

type Result struct {
	BidID  int
	Target float64
	Input  float64
	Output float64
}

var (
	results []Result
	mu      sync.Mutex
)

func runSimulation(wg *sync.WaitGroup, task SimulationTask) {
	defer wg.Done()

	pid := PIDController{Kp: 0.01, Ki: 0.005, Kd: 0.001}
	currentU := 0.0
	dt := 0.1
	maxSteps := 500 // Increased for higher precision

	for i := 0; i < maxSteps; i++ {
		y := System(currentU)
		error := task.Target - y

		if math.Abs(error) < 1e-6 {
			mu.Lock()
			results = append(results, Result{task.BidID, task.Target, currentU, y})
			mu.Unlock()
			return
		}

		deltaU := pid.Calculate(task.Target, y, dt)
		currentU += deltaU
	}

	mu.Lock()
	results = append(results, Result{task.BidID, task.Target, currentU, System(currentU)})
	mu.Unlock()
}

func main() {
	fmt.Println("EcoSphere - Atsushi Shibata: Break-even PID Calculation (100m units)")
	fmt.Println("System: Grand Strategy Map (Output = Input * 10 + 1)")
	fmt.Println("Running competitive bidding simulations using Goroutines...\n")

	targets := []float64{100.0, 150.0, 200.0, 250.0, 300.0}

	var wg sync.WaitGroup
	for i, target := range targets {
		wg.Add(1)
		task := SimulationTask{i + 1, target}
		go runSimulation(&wg, task)
	}

	wg.Wait()

	fmt.Println("---------------------------------------------------------------")
	fmt.Printf("%-10s | %-15s | %-15s | %-15s\n", "Bid ID", "Target (100m)", "Input (PID)", "Output (Actual)")
	fmt.Println("---------------------------------------------------------------")
	for _, res := range results {
		fmt.Printf("%-10d | %-15.2f | %-15.6f | %-15.6f\n", res.BidID, res.Target, res.Input, res.Output)
	}
	fmt.Println("---------------------------------------------------------------")
	fmt.Println("All simulations completed.")
}
